### PR TITLE
Don't create new method with public visibility scope

### PR DESCRIPTION
**Pull request by tjoyal**
Proposed extra visibility tune-up on top of #65 

I do not think there is any reasons for dynamically created methods `method_name_without_statsd` and `method_name_with_statsd` to be generated in the public scope.

At first I tried to bing the scope to the caller scope, but turns out some projects are already using this gem in their initialization this way, leading to a public scope which I think is not right.
```ruby
Object.extend StatsD::Instrument
Object.statsd_measure :call, 'Object.call'
```

Also, found and fix an issue where removing a StatsD would not always restore the appropriate scope. (Should I make a PR just for that?)

Please review @gmalette @wvanbergen

### CODE REVIEW

File: lib/statsd/instrument.rb [hunk 1]
```diff
@@ -210,28 +210,39 @@ def add_to_method(method, name, action, &block)
...
 
-      method_scope = case
-      when private_method_defined?(method)
-        :private
-      when protected_method_defined?(method)
-        :protected
-      else
-        :public
-      end
+      method_scope = visibility_scope_for(method)
```
Comment (gmalette) [inline lib/statsd/instrument.rb]: `s/visibility_scope_for/method_visibility`

**Comment by gmalette** [comment]
:+1:

**Comment by wvanbergen** [comment]
What is the visibility of the alias methods? (e.g. `method_without_statsd`)

**Comment by tjoyal** [comment]
> What is the visibility of the alias methods? (e.g. `method_without_statsd`)

My tests (and the code I'm trying to fix with this PR) showed that even if the helper method (eg.: `statsd_measure`) is called from a `private` scope, since the helper itself is `public` the `define_method` in it is run within that scope.

I'm fairly confident but do not have past experience with alias methods and scopes. Would you prefer to add an additional person to the discussion?

**Comment by tjoyal** [comment]
ping :)

**Comment by wvanbergen** [comment]
I am OK with this but I'd like to ping somebody with Rubt expertise on this one.

**Comment by wvanbergen** [comment]
Fixes https://github.com/Shopify/cloud/issues/175

**Comment by tjoyal** [comment]
@sgnr can I have your Rubt expertise on this please :smile_cat:

**Comment by sgnr** [comment]
If I understand correctly, you question is: which visibility is given to method created through `define_method`?

**Comment by sgnr** [comment]
If so, it is always `public`: https://github.com/ruby/ruby/blob/d9ac64fdc79ec2aefe5ed21b43b151caa296c5e3/class.c#L1487..L1491

**Comment by tjoyal** [comment]
But are you ok with the PR itself? :)

I do not really want them to be public

**Comment by sgnr** [comment]
Yes that looks sound :raised_hands:

**Pull request merged**